ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+>   the tuple `(freeze_asset, asset_frozen)` (for instance, by setting a `frozen`
+>   flag in _global_ storage).
+> - Upon a call to `account_freeze` the controlling Smart Contract SHOULD persist
+>   the tuple `(freeze_asset, freeze_account, asset_frozen)` (for instance by
+>   setting a `frozen` flag in the _local_ storage of the `freeze_account`). See
+>   the [security considerations section](#security-considerations) for how to
+>   ensure that Smart ASA holders cannot reset their `frozen` flag by clearing out
+>   their state at the controlling Smart Contract.
+
+```json
+[
+  {
+    "name": "get_asset_is_frozen",
+    "args": [{ "type": "asset", "name": "freeze_asset" }],
+    "returns": { "type": "bool" }
+  },
+  {
+    "name": "get_account_is_frozen",
+    "args": [
+      { "type": "asset", "name": "freeze_asset" },
+      { "type": "account", "name": "freeze_account" }
+    ],
+    "returns": { "type": "bool" }
+  }
+]
+```
+
+The value return by `get_asset_is_frozen` (respectively,
+`get_account_is_frozen`) tells whether any account (respectively
+`freeze_account`) can transfer or receive `freeze_asset`. A `false` value
+indicates that the transfer will be rejected.
+
+> Upon a call to `get_asset_is_frozen`, a reference implementation SHOULD
+> retrieve the tuple `(freeze_asset, asset_frozen)` as stored on `asset_freeze`
+> and return the value corresponding to `asset_frozen`.
+> Upon a call to `get_account_is_frozen`, a reference implementation SHOULD
+> retrieve the tuple `(freeze_asset, freeze_account, asset_frozen)` as
+> stored on `account_freeze` and return the value corresponding to
+> `asset_frozen`.
+
+#### Asset Destroy
+
+```json
+{
+  "name": "asset_destroy",
+  "args": [{ "type": "asset", "name": "destroy_asset" }],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_destroy` destroys a Smart ASA.
+
+> Upon a call to `asset_destroy`, a reference implementation SHOULD:
+>
+> - Fail if `destroy_asset` does not correspond to an ASA controlled by this smart
+>   contract.
+>
+> It is RECOMMENDED for calls to this method to be permissioned (see
+> `asset_create`).
+>
+> The controlling Smart Contract SHOULD perform an asset destroy operation on
+> the ASA with ID `destroy_asset`. The operation will fail if the asset is still
+> in circulation.
+
+#### Circulating Supply
+
+```json
+{
+  "name": "circulating_supply",
+  "args": [{ "type": "asset", "name": "asset" }],
+  "returns": { "type": "uint64" }
+}
+```
+
+Calling `circulating_supply` returns the circulating supply of a Smart ASA.
+
+> Upon a call to `circulating_supply`, a reference implementation SHOULD:
+>
+> - Fail if `asset` does not correspond to an ASA controlled by this smart
+>   contract.
+> - Return the circulating supply of `asset`, defined by the difference between
+>   the ASA `total` and the balance held by its `reserve_addr` (see [Asset
+>   Creation](#asset_creation)).
+
+#### Getters
+
+```json
```

Comment:
Should there be a getter that returns all of them in one payload too for convenience, or is the intention these are always going to be called from within smart contracts?

Making an ABI call to a `get_*` method is stored on the blockchain, you can't do that as an offline operation so it costs ALGO to execute a get?

Related: has any thought been given to how smart ASAs can be indexed without clogging the blockchain with heaps of get calls?

I guess you could index the app calls and make assumptions that the values passed in are persisted if the call succeeded. Not as clean as being able to call algod api to return current state of param values for an ASA (for no cost) since you now have to have an indexer. I guess the Algorand indexer could be modified to detect and support indexing smart ASAs though?

Reply:
I think the get_* methods should be marked as `read-only` as in ARC22 https://github.com/algorandfoundation/ARCs/pull/79

If they are, we should be able to call against dryrun or the in progress `simulate` endpoint which should be at no cost https://github.com/algorand/go-algorand/pull/4322

Reply:
@aldur reminded me that you can actually issue a dry run to call a get operation. That's pretty cool, but should definitely be mentioned / recommended (and definitely means there should be a `get_all_fields` method so you can achieve it in a single HTTP request to algod).

Reply:
(sorry, missed your comment there @barnjamin - great callout!)

Reply:
(and for clarity - agree they should be marked readonly, but also I think we need an endpoint that returns everything so dApps are more performant).

Reply:
Marked read-only methods, will also add a unique endpoint to get the config 👍

Reply:
Added a unique endpoint as well (replacing all the others to prevent redundancy).